File: Hash Table/group_anagrams.py
def groupAnagrams(strs):
    # create a dictionary to store the groups of anagrams
    groups = {}
    
    # iterate through each string in the list
    for s in strs:
        # sort the characters of the string to get its unique signature
        # for example, "eat" and "ate" both have the same signature "aet"
        sig = ''.join(sorted(s))
        
        # check if the signature is already in the dictionary
        # if not, add a new key-value pair with an empty list as the value
        if sig not in groups:
            groups[sig] = []
        
        # append the current string to the list of its corresponding signature
        groups[sig].append(s)
    
    # return the values of the dictionary as a list of lists
    return list(groups.values())

File: Hash Table/group_anagrams.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> mp; // create an unordered map to store anagrams

    // iterate through each string in the input vector
    for (string s : strs) {
        string t = s; // create a copy of the current string
        sort(t.begin(), t.end()); // sort the characters in the copy
        mp[t].push_back(s); // add the original string to the corresponding anagram group
    }

    vector<vector<string>> res; // create a vector to store the anagram groups
    for (auto p : mp) { // iterate through the unordered map
        res.push_back(p.second); // add the anagram group to the result vector
    }

    return res; // return the result vector containing all the anagram groups
}

